packages/transaction-controller/src/TransactionController.ts

```diff
@@ -1292,6 +1136,194 @@ export class TransactionController extends BaseController<
     return latestIncomingTxBlockNumber;
   }
 
+  private async processApproval(
+    transactionMeta: TransactionMeta,
+  ): Promise<string> {
+    const transactionId = transactionMeta.id;
+    let rejected = false;
+    let resultCallbacks: AcceptResultCallbacks | undefined;
+
+    try {
+      const acceptResult = await this.requestApproval(transactionMeta);
+      resultCallbacks = acceptResult.resultCallbacks;
+
+      const updatedMeta = this.getTransaction(transactionId);
+
+      const isCompleted =
+        updatedMeta && this.isLocalFinalState(updatedMeta.status);
+
+      if (updatedMeta && !isCompleted) {
+        await this.approveTransaction(transactionId);
+      }
+    } catch (error: any) {
+      const updatedMeta = this.getTransaction(transactionId);
+
+      const isCompleted =
+        updatedMeta && this.isLocalFinalState(updatedMeta.status);
+
+      if (updatedMeta && !isCompleted) {
+        if (error.code === errorCodes.provider.userRejectedRequest) {
+          this.cancelTransaction(transactionId);
+          rejected = true;
+        } else {
+          this.failTransaction(updatedMeta, error);
+        }
+      }
+    }
+
+    if (rejected) {
+      throw ethErrors.provider.userRejectedRequest(
+        'User rejected the transaction',
+      );
+    }
+
+    const finalMeta = this.getTransaction(transactionId);
+
+    if (finalMeta && finalMeta.status === TransactionStatus.failed) {
```

Comment:
For consistency, can we change this if and the one that compares with `submitted` to use optional chaining like in the `cancelled` check?

nit: I'm not a fan of switch statements, but here it might look better just to have one with options for `failed`, `cancelled` and `submitted`, and then the unknown problem one as a default.

Reply:
Done.